### PR TITLE
documents: rewrite public holdings/items view in Angular

### PR DESCRIPTION
**Pull request by Garfield-fr**
* Implements the holdings section of the document detailed
view of the public interface to improve the user experience when
loading holdings with lots of items. With the JINJA templates, the
performance is very bad. Angular allows to lazy load data and will
make easier to add dynamic interaction between the user and the interface.
* Adds pagination on holdings of the professional interface.

* Closes rero/rero-ils#1401, closes rero/rero-ils#1577

Co-Authored-by: Bertrand Zuchuat <bertrand.zuchuat@rero.ch>

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Cypress tests successful?

### CODE REVIEW

**Review by iGormilhit — APPROVED** [review]
Commit message approved.

File: projects/public-holdings-items/.browserslistrc [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+# https://github.com/browserslist/browserslist#queries
+
+# You can see what browsers were selected by your queries by running:
+#   npx browserslist
+
+> 0.5%
+last 2 versions
+Firefox ESR
+not dead
+not IE 9-11 # For IE 9-11 support, remove 'not'.
```
Comment (zannkukai) [inline projects/public-holdings-items/.browserslistrc:12]: What's the purpose of this file ?

File: projects/public-search/src/app/api/holdings.service.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+   * @param viewcode - string
+   * @return Observable
+   */
+  getHoldingsByDocumentPidAndViewcode(
+    documentPid: string, viewcode: string, page: number, itemsPerPage: number = 5): Observable<QueryResponse> {
+    return this._recordService
+    .getRecords('holdings', `document.pid:${documentPid}`, page, itemsPerPage, undefined, { view: viewcode }, this._headers)
+    .pipe(map((response: Record) => {
+        return response.hits;
+    }));
```
Comment (zannkukai) [inline projects/public-search/src/app/api/holdings.service.ts:56]: A little bit more indentation ? :-)
```
return this._recordService
    .getRecords('holdings', `document.pid:${documentPid}`, page, itemsPerPage, undefined, { view: viewcode }, this._headers)
    .pipe(map((response: Record) => response.hits));
```

File: projects/public-search/src/app/api/item.service.ts [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+   * @param page - number
+   * @param itemsPerPage - number
+   * @return Observable
+   */
+  getItemsByHoldingsPidAndViewcode(
+    holdingsPid: string, viewcode: string, page: number, itemsPerPage: number = 5): Observable<QueryResponse> {
+    return this._recordService
+      .getRecords(
+        'items', `holding.pid:${holdingsPid}`, page, itemsPerPage, undefined,
+        { view: viewcode }, this._headers, 'enumeration_chronology'
```
Comment (zannkukai) [inline projects/public-search/src/app/api/item.service.ts:60]: is `enumeration_chronology` sort field is defined into the backend config ?

File: projects/public-search/src/app/api/item.service.ts [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+   * @return Observable
+   */
+  getItemsByHoldingsPids(
+    holdingsPids: string[], viewcode: string, page: number, itemsPerPage: number = 5): Observable<QueryResponse> {
+    const query = 'holding.pid:' + holdingsPids.join(' OR holding.pid:');
+    return this._recordService
+      .getRecords('items', query, page, itemsPerPage, undefined, { view: viewcode }, this._headers)
+      .pipe(map((response: Record) => {
+        return response.hits;
+      }));
```
Comment (zannkukai) [inline projects/public-search/src/app/api/item.service.ts:99]: same

File: projects/public-search/src/app/document-detail/holdings/holding/holding.component.html [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+  You should have received a copy of the GNU Affero General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+-->
+<div id="holding-{{ holding.metadata.pid }}" class="card-header p-2">
+  <div class="row">
+    <div id="holding-location-{{ holding.metadata.pid }}" class="col-sm-4">
+      <a href="#" class="collapse-link mr-2" data-toggle="collapse" aria-expanded="false" (click)="isCollapsed = !isCollapsed">
+        <i [ngClass]="{ 'fa-caret-down': !isCollapsed, 'fa-caret-right': isCollapsed }" class="fa fa-lg" aria-hidden="true"></i>
+      </a>
+      {{ holding.metadata.library.name }}: {{ holding.metadata.location.name }}
```
Comment (zannkukai) [inline projects/public-search/src/app/document-detail/holdings/holding/holding.component.html:23]: don't know how to manage spaces before/after semi-colon by language.... discussion is now open :D

File: projects/public-search/src/app/document-detail/holdings/holding/holding.component.html [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    </div>
+    <div id="holding-category-name-{{ holding.metadata.pid }}" class="col-sm-3">
+      {{ holding.metadata.circulation_category.name }}
+    </div>
+    <div class="col-sm-2">
+      <span [hidden]="itemsCount == 0">
+        {{ itemsCount }} {{ (itemsCount > 1 ? 'items' : 'item') | translate }}
```
Comment (zannkukai) [inline projects/public-search/src/app/document-detail/holdings/holding/holding.component.html:30]: is this construction allows to extract the strings ?

**Review by zannkukai — APPROVED** [review]

File: projects/public-search/src/app/api/holdings.service.ts (hunk 2 above)
Comment (Garfield-fr) [reply projects/public-search/src/app/api/holdings.service.ts:56]: done

File: projects/public-search/src/app/api/item.service.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+   * @return Observable
+   */
+  getItemsByHoldingsPidAndViewcode(
+    holdingsPid: string, viewcode: string, page: number, itemsPerPage: number = 5): Observable<QueryResponse> {
+    return this._recordService
+      .getRecords(
+        'items', `holding.pid:${holdingsPid}`, page, itemsPerPage, undefined,
+        { view: viewcode }, this._headers, 'enumeration_chronology'
+      ).pipe(map((response: Record) => {
+        return response.hits;
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/public-search/src/app/api/item.service.ts:62]: done

File: projects/public-search/src/app/api/item.service.ts (hunk 3 above)
Comment (Garfield-fr) [reply projects/public-search/src/app/api/item.service.ts:60]: yes, here: https://github.com/rero/rero-ils/pull/1611/files#diff-6be28a31924cc25a9c440ab46eedfa749b3b94ec7a6c5bb3021f76530ac95e4aR2015

File: projects/public-search/src/app/document-detail/holdings/holding/holding.component.html (hunk 6 above)
Comment (Garfield-fr) [reply projects/public-search/src/app/document-detail/holdings/holding/holding.component.html:30]: `{{ itemsCount | i18nPlural: { '<=1': 'item', 'other': 'items' } | translate }}`

File: projects/public-search/src/app/document-detail/request/pickup-location/pickup-location.component.html [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+-->
+<ng-container *ngIf="fields.length > 0 && showForm">
+  <div id="pickup-location-{{ item.metadata.pid }}" class="row pl-4 col-12">
+    <dt class="mb-0 col-lg-2 col-sm-3">&nbsp;</dt>
+    <dd class="mb-0 col-lg-10 col-sm-9">
+      <form class="my-2" [formGroup]="form" (ngSubmit)="submit()">
+        <formly-form [model]="model" [fields]="fields" [form]="form"></formly-form>
+        <button type="submit" class="btn btn-sm btn-primary" [disabled]="!form.valid || requestInProgress">
+          <ng-container *ngIf="!requestInProgress; else inProcess">
+            {{ 'Confirm you request' | translate }}
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/public-search/src/app/document-detail/request/pickup-location/pickup-location.component.html:25]: done

File: projects/public-search/src/app/document-detail/request/request.component.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+    return this._userService.user !== undefined;
+  }
+
+  /**
+   * Constructor
+   * @param _ItemService - ItemService
+   * @param _userService - UserService
+   */
+  constructor(
+    private _ItemService: ItemService,
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/public-search/src/app/document-detail/request/request.component.ts:53]: done

File: projects/public-search/src/app/user.service.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+   * Constructor
+   * @param _loggedUserService - LoggedUserService
+   */
+  constructor(private _loggedUserService: LoggedUserService) {}
+
+  init() {
+    this._loggedUserService.onLoggedUserLoaded$
+      .pipe(map(data => {
+        return data.metadata ? data.metadata : undefined;
+      })).subscribe(user => this.currentUser = user);
```
[the remark this answers is not in the document]
Comment (Garfield-fr) [reply projects/public-search/src/app/user.service.ts:44]: done

**Review by zannkukai — APPROVED** [review]

**Review by AoNoOokami — APPROVED** [review]

**Pull request merged**